### ISSUE TITLE
Add Vim support and improve Emacs support

### ISSUE THREAD

**Issue by lexxish** [dictation-toolbox/Caster#534]
This will require a significant amount of refactoring due to how Vim handles modes. Exact method TBD but prefer to keep Vim in command mode or allow either command or insert mode.

**Comment by kendonB** [comment]
I think the simple answer is correct for vim, which is to have voice commands for all the common keystroke combinations. e.g. something like
```python
"insert mode":
    R(Key("i")),
"save and close":
    R(Key("escape, colon, x, enter")),
```
@lexxish what complications do you envision? For every command that requires command mode, prefix with `escape`. For caster commands that require you to already be in insert mode it's fine to assume the user already is I think.

**Comment by lexxish** [comment]
I was considering assuming the user spends most the time in command mode which means Caster commands would have to be overridden. Also, after doing some testing it looks like some Caster commands such as "shin" and "queue" don't even work in insert mode.

**Comment by RyanQuey** [comment]
I'm having a hard time using vim with caster also, it looks like commands that should be used in insert mode that use `Text("some text here")` don't work well, though `Key("some key")` works fine.  

Pretty much renders rules for programming (eg the html rule) useless

**Comment by LexiconCode** [comment]
> 
> I'm having a hard time using vim with caster also, it looks like commands that should be used in insert mode that use `Text("some text here")` don't work well, though `Key("some key")` works fine.
> 
> Pretty much renders rules for programming (eg the html rule) useless

It would be nice to have a proper grammar of the modes and so forth. Until that time consider the following.

`Key` is slow. However you could try utilizing the `Paste` action instead in `html`. To make it simple without heavy edits to existing grammar use `from dragonfly import Paste as Text`

**Comment by lexxish** [comment]
I have basic vim mode tracking and commands here https://gist.github.com/lexxish/1a9582f175119f36a508282a81761d2a

I haven't had a chance to merge them into caster yet. You'd want the files vim_helper, vim_mode, vim_rule and you'd replace InheritedRule with MappingRule

**Comment by LexiconCode** [comment]
Awesome and of course right after I removed the WIP tag :)

**Comment by RyanQuey** [comment]
This is great! Thanks guys, this is super helpful!

**Comment by LexiconCode** [comment]
> 
> 
> I have basic vim mode tracking and commands here [gist.github.com/lexxish/1a9582f175119f36a508282a81761d2a](https://gist.github.com/lexxish/1a9582f175119f36a508282a81761d2a)
> 
> I haven't had a chance to merge them into caster yet. You'd want the files vim_helper, vim_mode, vim_rule and you'd replace InheritedRule with MappingRule

@lexxish function context https://github.com/dictation-toolbox/Caster/issues/747 will go a long way to make modes much easier to implement.

**Comment by LexiconCode** [comment]
All right @lexxish FuncContext is now supported in Caster.

FuncContext a dragonfly class that evaluates a given function with optional parameters for window title/executable, whose return value is interpreted as a bool, determining whether the context is active.

**Comment by lubyrex** [comment]
Hi, I was wondering if there is any update on this.  Or if anyone knows of good software on speech recognition for vim. Thanks.

**Comment by Talos878** [comment]
Is there documentation for a standardized required set of commands for an IDE  to be accepted? I use vim, and wood like to help if possible.

**Comment by LexiconCode** [comment]
https://www.vim.org/

I'm willing to do this in collaboration with help. Someone could create the individual grammars consisting of one mode per file. Then I would be willing to do the plumbing to create the modes using function context.

My recommendation would to develop a grammar for a specific vim editor 1st. After that we can figure out how generalizable it is to other vim like Programs like vs code in vim mode.

Vim Cheat Sheet  https://vim.rtorr.com/

Using the above cheat cheat we can create different modes function context. A final pass would probably need to be done to review the commands.

**Comment by Talos878** [comment]
Using contexts, is it possible for a rule to know what mode it is in, and if a rule is for another mode switch to the proper mode and run the command from the other file?

**Comment by LexiconCode** [comment]
Yes, function context gives us that functionality.  What function context does is evaluate a function to true or false every utterance.  If true, then the context and therefore rule becomes active.  This is what I did with the mouse grids utilizing one function.

**Comment by LexiconCode** [comment]
In a couple of days here, I'll come up with a minimum implementation that you guys can use to expand.

**Comment by LexiconCode** [comment]
@Talos878 @lubyrex
Here you go [vim_rules.zip](https://github.com/dictation-toolbox/Caster/files/7923822/vim_rules.zip). a minimum implementation that you guys can use to expand.
 

Test from Run_Caster_Test_Engine.bat
```
enable vim
ex # or mode name
test mode # prints active mode name eg "ex-mode"
ex-mode
```

**Comment by Talos878** [comment]
Thanks.
